File: src/miscellaneous/binarySearch/BinarySearch.java
package miscellaneous.binarySearch;

/**
 * 704. {@link <a href="https://leetcode.com/problems/binary-search/">Binary Search</a>}
 * <p>
 * Given an array of integers nums which is sorted in ascending order, and an integer target,
 * write a function to search target in nums. If target exists, then return its index. Otherwise, return -1.
 * <p>
 * You must write an algorithm with O(log n) runtime complexity.
 * <p>
 * Example 1:
 * <p>
 * Input: nums = [-1,0,3,5,9,12], target = 9
 * Output: 4
 * Explanation: 9 exists in nums and its index is 4
 * </p>
 * Example 2:
 * <p>
 * Input: nums = [-1,0,3,5,9,12], target = 2
 * Output: -1
 * Explanation: 2 does not exist in nums so return -1
 * <p>
 * Constraints:
 * <p>
 * 1 <= nums.length <= 104
 * -104 < nums[i], target < 104
 * All the integers in nums are unique.
 * nums is sorted in ascending order.
 */

public class BinarySearch {

    public static void main(String[] args) {
        int[] array = {2, 4, 6, 8, 9, 12};
        System.out.print("Index = " + search(array, 12));
    }

    public static int search(int[] nums, int target) {
        int start = 0;
        int end = nums.length - 1;
        while (start <= end) {
            int mid = (end - start) / 2 + start;
            if (nums[mid] == target) {
                return mid;
            } else if (target < nums[mid]) {
                end = mid - 1;

            } else if (target > nums[mid]) {
                start = mid + 1;
            }
        }
        return -1;
    }

    public int recursiveSearch(int[] nums, int target, int start, int end) {
        int mid = (end - start) / 2 + start;
        if (start > end) {
            return -1;
        } else if (nums[mid] == target) {
            return mid;
        } else if (target > nums[mid]) {
            return recursiveSearch(nums, target, mid + 1, end);
        } else if (target < nums[mid]) {
            return recursiveSearch(nums, target, start, end - 1);
        } else {
            return -1;
        }
    }
}

File: src/searching/BinarySearch.kt
package searching

/**
 * Find the index of a given number n in a given sorted array.
 * In case the given number n is not available in the given array return -1
 *
 * I/P: [2, 4, 5, 7, 9], 7
 * O/P: 3
 *
 * I/P: [0, 7, 9, 11], 7
 * O/P: 1
 *
 * I/P: [3, 5, 6, 11], 7
 * O/P: -1
 *
 */
val sortedArray = arrayOf(2, 4, 5, 7, 9)
val sorted = arrayOf(2)
fun main() {
    println(binarySearchIterative(sortedArray, 7))
    println(binarySearch(sorted, 0, sorted.size - 1, 2))
}

fun binarySearchIterative(array: Array<Int>, n: Int): Int {
    var start = 0
    var end = array.size - 1

    while (start <= end) {
        val mid = start + (end - start) / 2
        if (n == array[mid]) {
            return mid
        }
        if (n > array[mid]) {
            start = mid + 1
        } else {
            end = mid - 1
        }
    }
    return -1
}


fun binarySearch(array: Array<Int>, start: Int, end: Int, n: Int): Int {
    if (start > end) {
        return -1
    }
    var mid = start + (end - start) / 2
    return if (array[mid] == n) {
        mid
    } else if (n > array[mid]) {
        binarySearch(array, mid + 1, end, n)
    } else {
        binarySearch(array, start, mid - 1, n)
    }
}